src/bootstrap/config.rs

```diff
@@ -805,7 +804,41 @@ impl Config {
         config.llvm_profile_use = flags.llvm_profile_use;
         config.llvm_profile_generate = flags.llvm_profile_generate;
 
+        // Infer the rest of the configuration.
+
+        // Infer the source directory. This is non-trivial because we want to support a downloaded bootstrap binary,
+        // running on a completely machine from where it was compiled.
+        let mut cmd = Command::new("git");
+        // NOTE: we cannot support running from outside the repository because the only path we have available
+        // is set at compile time, which can be wrong if bootstrap was downloaded from source.
+        // We still support running outside the repository if we find we aren't in a git directory.
+        cmd.arg("rev-parse").arg("--show-toplevel");
+        // Discard stderr because we expect this to fail when building from a tarball.
+        let output = cmd
+            .stderr(std::process::Stdio::null())
+            .output()
+            .ok()
+            .and_then(|output| if output.status.success() { Some(output) } else { None });
+        if let Some(output) = output {
+            let git_root = String::from_utf8(output.stdout).unwrap();
+            config.src = PathBuf::from(git_root.trim().to_owned())
```

Comment:
I've tried running this; it seems to replace `"C:\\Users\\bruno\\Rust\\rustpr\\rust"` with `"C:/Users/bruno/Rust/rustpr/rust"`, which then causes a bunch of failing mir-opt tests.

Reply:
We worked through this together and found that `crate::util::absolute(std::env::current_exe().unwrap().join("../../../.."))` does the same thing and also works on Windows, as well as supporting out-of-tree builds when building from a tarball :)

Reply:
err no it doesn't actually do that last thing, the build directory need not be inside the source directory :( so we're going to hack this by transforming `\` to `/ when setting `CFG_VIRTUAL_RUST_SOURCE_BASE_DIR`.

Reply:
I think the hacky solution for now is to just trim this manually. e.g. 
```suggestion
            let git_root = PathBuf::from(git_root.trim()).canonicalize().unwrap();
            let git_root = git_root.to_str().unwrap();

            // Bootstrap is quite bad at handling /? in front of paths
            config.src = match git_root.strip_prefix("\\\\?\\") {
                Some(p) => PathBuf::from(p),
                None => PathBuf::from(git_root),
            };
```

Reply:
god I hate this lol

yeah that seems reasonable as a stop-gap

Reply:
Committed. Sorry wasnt able to tes it myself lets see if ci passes.